### DISCUSSION TITLE
Support "plain" code_challenge_method

### DISCUSSION THREAD

**Discussion by kotx** [nextauthjs/next-auth#4459] [Help] [answered]
### Description 📓

https://next-auth.js.org/errors#pkce_error:
> The default `code_challenge_method` is `"S256"`. This is currently not configurable to `"plain"`, as it is not recommended, and in most cases, it is only supported for backward compatibility.

I am trying to implement and submit a builtin provider for MyAnimeList, but it only allows a `plain` `code_challenge_method`.

### How to reproduce ☕️

Try to use a provider that only supports plain code_challenge_method, [like MyAnimeList](https://myanimelist.net/blog.php?eid=835707).
Support something like a `codeChallengeMethod` option in `Provider` or `AuthorizationEndpointHandler`.

### Contributing 🙌🏽

Yes, I am willing to help implement this feature in a PR

**Comment by ThangHuuVu** [comment, marked as answer]
Unfortunately, I don't think it's possible to support this in configuration. The documentation made it very clear: we don't want to support `plain` because of the security implication. You'll need to maintain your fork separately in this case.

**Reply by kotx** [reply]
I don't think documentation is very clear. It's not good to not support something in favor of backwards compatibility. However,  [RFC7636](https://www.rfc-editor.org/rfc/rfc7636#section-4.2) says:
> If the client is capable of using "S256", it MUST use "S256", as
   "S256" is Mandatory To Implement (MTI) on the server.  Clients are
   permitted to use "plain" only if they cannot support "S256" for some
   technical reason and know via out-of-band configuration that the
   server supports "plain".

So not supporting this makes sense. Is it alright if I update the documentation to point to this? It provides an actual reason as opposed to just "it's not recommended" (by who?) and "it's mostly only supported for backwards compatibility" (it doesn't seem to be? those clients would be violating the specification anyways). I can open a PR for it, if you'd like.

**Reply by ThangHuuVu** [reply]
I guess you can try to submit a PR to improve it. The [docs](https://next-auth.js.org/errors#pkce_error) already mentioned [RFC7636](https://www.rfc-editor.org/rfc/rfc7636#section-4.2) 
> The provider you tried to use failed when setting [PKCE or Proof Key for Code Exchange](https://tools.ietf.org/html/rfc7636#section-4.2).

**Reply by Pelps12** [reply]
Was a PR ever submitted for it?

**Reply by kotx** [reply]
@Pelps12 Yup, https://github.com/nextauthjs/next-auth/pull/4464

**Reply by Pelps12** [reply]
How did you work around the MyAnimeList API. It feels very legacy

**Reply by kotx** [reply]
> How did you work around the MyAnimeList API. It feels very legacy

Sadly I didn't :P
If I had to guess you would need to implement it manually. Unsure if that's doable with NextAuth with a custom thing or whatever
